### PR TITLE
Support propagation the authentication context of a Manual Judgment s…

### PR DESCRIPTION
**Pull request by ajordens**
…tage's approver

 Handles the scenario where a pipeline was triggered by credentials that did not allow
 access to a particular account.

 In such a situation it would be nice if the pipeline could continue to run under the context
 of the individual that approved a manual judgment step.

### CODE REVIEW

**Comment by ajordens** [comment]
This turned into a bit more than a "quick" hack.

Essentially this allows a pipeline to continue under the authentication context of the individual that approved a manual judgment step. 

Is this a terrible idea?

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/jedis/JedisExecutionRepository.groovy [hunk 1]
```diff
@@ -467,51 +471,6 @@ class JedisExecutionRepository implements ExecutionRepository {
     }
   }
 
-  @Deprecated
-  @CompileDynamic
-  private <T extends Execution> T sortStages(JedisCommands jedis, T execution, Class<T> type) {
```
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/jedis/JedisExecutionRepository.groovy]: @robfletcher These methods weren't being called so took the liberty of removing

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 2]
```diff
@@ -54,6 +56,20 @@ class ManualJudgmentStage extends LinearStage implements RestartableStage {
...
+    def stageData = stage.mapTo(StageData)
+    if (stageData.state != StageData.State.CONTINUE || !stage.lastModified?.user || !stageData.propagateAuthenticationContext) {
+      return Optional.empty()
+    }
+
+    def user = new User()
+    user.setAllowedAccounts(stage.lastModified.allowedAccounts)
+    user.setUsername(stage.lastModified.user)
+    user.setEmail(stage.lastModified.user)
+    return Optional.of(user)
```
Comment (ttomsu) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: `return Optional.of(user.asImmutable())`

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy [hunk 3]
```diff
@@ -54,6 +56,20 @@ class ManualJudgmentStage extends LinearStage implements RestartableStage {
...
 
+  @Override
+  Optional<User> authenticatedUser(Stage stage) {
+    def stageData = stage.mapTo(StageData)
+    if (stageData.state != StageData.State.CONTINUE || !stage.lastModified?.user || !stageData.propagateAuthenticationContext) {
+      return Optional.empty()
+    }
+
+    def user = new User()
+    user.setAllowedAccounts(stage.lastModified.allowedAccounts)
```
Comment (ttomsu) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: We'll likely need to yank this out once Fiat comes online.

**Comment by ttomsu** [comment]
Interesting idea. Though, if someone without access presses the "Go" button, the whole pipeline fails, right? And without controls on who else can see your applications pipelines, lots of unauthorized people could press the buttons of paused pipelines.

**Comment by ajordens** [comment]
They could click the pipeline and it would fail ... that's the behavior we currently have (which is kind of annoying with locked accounts).

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 3 above)
Comment (ajordens) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: Certainly ... with fiat checks happening in clouddriver we would only need to inject the user and not the accounts.

File: orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy (hunk 2 above)
Comment (ajordens) [inline orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/pipeline/ManualJudgmentStage.groovy]: Good catch!

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/batch/adapters/TaskTasklet.groovy [hunk 4]
```diff
@@ -173,9 +177,17 @@ class TaskTasklet implements Tasklet {
     }
 
     try {
-      def currentUser = new User(
+      // An AuthenticatedStage can override the default pipeline authentication credentials
+      def authenticatedUser = stageNavigator.findAll(stage, { Stage ancestorStage, StageBuilder stageBuilder ->
```
Comment (cfieber) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/batch/adapters/TaskTasklet.groovy]: its not clear what the ordering would be here (its it finding the stages by traversing backwards up the graph to the nearest ancestor?) If that is the case maybe adding a findNearestAncestor on stageNavigator would be useful

**Comment by cfieber** [comment]
I think this makes sense, definite improvement over the current state of things

File: orca-core/src/main/groovy/com/netflix/spinnaker/orca/batch/adapters/TaskTasklet.groovy (hunk 4 above)
Comment (ajordens) [inline orca-core/src/main/groovy/com/netflix/spinnaker/orca/batch/adapters/TaskTasklet.groovy]: Ya, that is what it's doing (traversing up the graph ... equivalent to a nearest ancestor).

**Pull request merged**